api/src/main/java/org/pmiops/workbench/initialcredits/InitialCreditsExpirationServiceImpl.java

```diff
@@ -17,4 +66,39 @@ public Optional<Timestamp> getCreditsExpiration(DbUser user) {
         // .filter(() -> institutionService.shouldBypassForCreditsExpiration(user))
         .map(DbUserInitialCreditsExpiration::getExpirationTime);
   }
+
+  private void checkCreditsExpirationForUser(DbUser user, Timestamp now) {
+    DbUserInitialCreditsExpiration userInitialCreditsExpiration =
+        user.getUserInitialCreditsExpiration();
+    if (null != userInitialCreditsExpiration
+        && !userInitialCreditsExpiration.isBypassed()
+        && userInitialCreditsExpiration
+            .getNotificationStatus()
+            .equals(InitialCreditExpirationNotificationStatus.NO_NOTIFICATION_SENT)
+        && !(userInitialCreditsExpiration.getExpirationTime().after(now))) {
+      expireBillingStatusForUserWorkspaces(user);
+      try {
+        mailService.alertUserInitialCreditsExpired(user);
+        userInitialCreditsExpiration.setNotificationStatus(
+            InitialCreditExpirationNotificationStatus.EXPIRATION_NOTIFICATION_SENT);
+        userInitialCreditsExpirationDao.save(userInitialCreditsExpiration);
+
+      } catch (MessagingException e) {
+        log.warning(
+            String.format(
+                "Failed to send initial credits expiration notification for user %s",
+                user.getUserId()));
+      }
+    }
+  }
+
+  private void expireBillingStatusForUserWorkspaces(DbUser user) {
+    workspaceDao.findAllByCreator(user).stream()
+        .filter(
+            ws ->
+                WorkspaceUtils.isFreeTier(
+                    ws.getBillingAccountName(), workbenchConfigProvider.get()))
+        .map(DbWorkspace::getWorkspaceId)
+        .forEach(id -> workspaceDao.updateBillingStatus(id, BillingStatus.EXPIRED));
```

Comment:
might be more performant to update all at once?

Reply:
Are you suggesting that I add method to the workspace dao that does something like:
```
  @Query(
      value =
          "UPDATE workspace"
              + " SET billingStatus = :billingStatus"
              + " WHERE workspace_id in (:workspacesIds)",
      nativeQuery = true)
```